### PR TITLE
feat(spells): support empty trigger configs [NET-316]

### PR DESCRIPTION
**Pull request by kmd-fl**
An empty trigger config allows:
1. not to subscribe a spell to any triggers on installation
2. unsubscribe a spell from its triggers

 Also, remove the update interface from the spell-event-bus crate since all it needs to do is expressed via subscribe/unsubscribe methods.

### CODE REVIEW

File: crates/particle-node-tests/tests/spells.rs [hunk 1]
```diff
@@ -946,3 +903,152 @@ fn spell_peer_id_test() {
 
     assert_eq!(result, scope_peer_id);
 }
+#[test]
+fn spell_update_config() {
+    let swarms = make_swarms(1);
+    //log_utils::enable_logs();
```
Comment (folex) [inline crates/particle-node-tests/tests/spells.rs:909]:
```suggestion
```

File: crates/particle-node-tests/tests/spells.rs [hunk 2]
```diff
@@ -946,3 +903,152 @@ fn spell_peer_id_test() {
...
+        r#"(seq
+            (seq
+                (call %init_peer_id% ("getDataSrv" "spell_id") [] spell_id)
+                (call %init_peer_id% (spell_id "list_pop_string") ["trigger_mailbox"] result)
+             )
+            (call "{}" ("return" "") [result])
+        )"#,
+        client.peer_id
+    );
+    // create periodic spell
```
Comment (kmd-fl) [inline crates/particle-node-tests/tests/spells.rs:924]:
```suggestion
```

File: crates/particle-node-tests/tests/spells.rs [hunk 3]
```diff
@@ -946,3 +903,152 @@ fn spell_peer_id_test() {
...
+        assert_matches!(info, TriggerInfo::Peer(p) if !p.connected, "spell must be triggered by the `disconnected` event");
+    } else {
+        panic!("wrong result from spell, expect trigger info with the `disconnected` event");
+    }
+}
+
+#[test]
+fn spell_update_config_stopped_spell() {
+    let swarms = make_swarms(1);
+    //log_utils::enable_logs();
```
Comment (folex) [inline crates/particle-node-tests/tests/spells.rs:990]:
```suggestion

```

File: crates/particle-node-tests/tests/spells.rs [hunk 4]
```diff
@@ -946,3 +903,152 @@ fn spell_peer_id_test() {
...
+    let mut client = ConnectedClient::connect_to(swarms[0].multiaddr.clone())
+        .wrap_err("connect client")
+        .unwrap();
+
+    // Create spell that do nothing.
```
Comment (folex) [inline crates/particle-node-tests/tests/spells.rs:995]:
```suggestion
    // Create spell that does nothing.
```

File: crates/spell-event-bus/src/bus.rs [hunk 5]
```diff
@@ -40,14 +40,10 @@ impl PeerEventSubscribers {
...
     /// Returns true if spell_id was removed from subscribers
-    fn remove(&mut self, spell_id: &SpellId) -> bool {
-        let mut was_removed = false;
+    fn remove(&mut self, spell_id: &SpellId) {
         for subscribers in self.subscribers.values_mut() {
-            let subscribers_len = subscribers.len();
             subscribers.retain(|sub_id| **sub_id != *spell_id);
-            was_removed |= subscribers_len != subscribers.len();
         }
-        was_removed
```
Comment (folex) [inline crates/spell-event-bus/src/bus.rs:50]: nice!

File: crates/spell-event-bus/src/bus.rs [hunk 6]
```diff
@@ -132,19 +128,10 @@ impl SubscribersState {
...
             .retain(|scheduled| *scheduled.data.id != *spell_id);
-        let new_len = self.scheduled.len();
-        let removed = self.subscribers.remove(spell_id);
-        prev_len != new_len || removed
-    }
-
-    fn update(&mut self, spell_id: &SpellId, config: &SpellTriggerConfigs) {
-        if self.unsubscribe(spell_id) {
-            self.subscribe(spell_id.clone(), config);
-        }
```
Comment (folex) [inline crates/spell-event-bus/src/bus.rs:147]: awesome ^__^

File: sorcerer/src/spells.rs [hunk 7]
```diff
@@ -88,19 +89,23 @@ pub(crate) async fn spell_install(
...
 
-        return Err(JError::new(format!(
-            "can't install a spell due to an internal error while subscribing to the triggers: {err}"
-        )));
+            return Err(JError::new(format!(
+                "can't install a spell due to an internal error while subscribing to the triggers: {err}"
+            )));
+        }
+    } else {
+        log::trace!("empty config given for spell {}", spell_id);
```
Comment (folex) [inline sorcerer/src/spells.rs:108]: can you include particle_id in this log please? I think one day we might want to see this message when debugging spell installation

**Review by folex — APPROVED** [review]

File: sorcerer/src/spells.rs [hunk 8]
```diff
@@ -171,16 +178,24 @@ pub(crate) async fn spell_update_config(
...
+        if let Some(config) = config {
+            // and if the config isn't empty, we subscribe it to the new one
+            spell_event_bus_api
+                .subscribe(spell_id.clone(), config.clone())
+                .await?;
+        }
+    };
+
+    if let Err(err) = result {
+        log::warn!("can't update a spell {spell_id} config via spell-event-bus-api: {err}",);
```
Comment (justprosh) [inline sorcerer/src/spells.rs:193]:
```suggestion
        log::warn!("can't update a spell {spell_id} config via spell-event-bus-api: {err}");
```

**Review by justprosh — APPROVED** [review]

**Review by folex — APPROVED** [review]

**Pull request merged**